contracts/src/RIP7755Source.sol

```diff
@@ -0,0 +1,271 @@
+    /// @param currentTimestamp The current block timestamp
+    /// @param expiry The timestamp at which the request expires
+    error CannotCancelRequestBeforeExpiry(uint256 currentTimestamp, uint256 expiry);
+
+    /// @notice Submits an RIP-7755 request for a cross chain call
+    ///
+    /// @param request A cross chain request structured as a `CrossChainRequest`
+    function requestCrossChainCall(CrossChainRequest memory request) external payable {
+        request.nonce = ++_nonce;
+        bool usingNativeCurrency = request.rewardAsset == _NATIVE_ASSET;
+        uint256 expectedValue = usingNativeCurrency ? request.rewardAmount : 0;
+
+        if (msg.value != expectedValue) {
+            revert InvalidValue(expectedValue, msg.value);
+        }
+
+        bytes32 requestHash = hashRequestMemory(request);
+        _requestMetadata[requestHash] = RequestMeta({
+            status: CrossChainCallStatus.Requested,
+            expiryTimestamp: uint40(block.timestamp + request.validDuration),
+            requester: msg.sender
+        });
+
+        emit CrossChainCallRequested(requestHash, request);
+
+        if (!usingNativeCurrency) {
+            _pullERC20({owner: msg.sender, asset: request.rewardAsset, amount: request.rewardAmount});
+        }
+    }
+
+    /// @notice To be called by a Filler that successfully submitted a cross chain request to the destination chain and
+    /// can prove it with a valid nested storage proof
+    ///
+    /// @param request A cross chain request structured as a `CrossChainRequest`
+    /// @param fillInfo The fill info that should be in storage in `RIP7755Verifier` on destination chain
+    /// @param storageProofData A storage proof that cryptographically verifies that `fillInfo` does, indeed, exist in
+    /// storage on the destination chain
+    /// @param payTo The address the Filler wants to receive the reward
+    function claimReward(
+        CrossChainRequest calldata request,
+        RIP7755Verifier.FulfillmentInfo calldata fillInfo,
+        bytes calldata storageProofData,
+        address payTo
+    ) external {
+        bytes32 requestHash = hashRequest(request);
+        bytes32 storageKey = keccak256(
+            abi.encodePacked(
+                requestHash,
+                uint256(0) // Must be at slot 0
+            )
+        );
+
+        _checkValidStatusForClaim(requestHash);
+
+        _validate(storageKey, fillInfo, request, storageProofData);
+        _requestMetadata[requestHash].status = CrossChainCallStatus.Completed;
+
+        _sendReward(request, payTo);
+    }
+
+    /// @notice Cancels a pending request that has expired
+    ///
+    /// @dev Can only be called if the request is in the `CrossChainCallStatus.Requested` state
+    ///
+    /// @param request A cross chain request structured as a `CrossChainRequest`
+    function cancelRequest(CrossChainRequest calldata request) external {
```

Comment:
should this be authed in some way?

Reply:
I thought about it but think adding auth will just be extra gas unnecessarily. We'll have timestamp-based protection that prevents cancellations before a request has expired. And a canceled request sends the reward funds back to the original requester (as opposed to providing a destination address with the cancellation txn)

Reply:
yes